sql/src/main/java/io/crate/planner/consumer/ManyTableConsumer.java

```diff
@@ -107,19 +123,27 @@ public Plan consume(AnalyzedRelation relation, ConsumerContext context) {
                                                            Set<? extends Set<QualifiedName>> implicitJoinedRelations,
                                                            List<JoinPair> joinPairs,
                                                            Collection<QualifiedName> preSorted) {
+        // All relations already sorted based the `ORDER BY` symbols
         if (relations.size() == preSorted.size()) {
             return preSorted;
         }
+
+        // Only 2 relations or the relations have no join conditions (explicit or implicit) between them
         if (relations.size() == 2 ||
             (joinPairs.isEmpty() && explicitJoinedRelations.isEmpty() && implicitJoinedRelations.isEmpty())) {
             LinkedHashSet<QualifiedName> qualifiedNames = new LinkedHashSet<>(preSorted);
             qualifiedNames.addAll(relations);
             return qualifiedNames;
         }
+
+        // If no `ORDER BY` present we have no preSort to follow so we return the relations in ordering
+        // obtained by the join conditions (explicit and/or implicit) between them
         if (preSorted.isEmpty()) {
             Set<QualifiedName> qualifiedNames = new LinkedHashSet<>(preSorted);
-            qualifiedNames.addAll(explicitJoinedRelations.stream().flatMap(Collection::stream).collect(Collectors.toList()));
-            qualifiedNames.addAll(implicitJoinedRelations.stream().flatMap(Collection::stream).collect(Collectors.toList()));
+            qualifiedNames.addAll(
+                explicitJoinedRelations.stream().flatMap(Collection::stream).collect(Collectors.toList()));
```

Comment:
here you create a list just to call addAll, you could use forEach(qualifiedNames::add)